cpp/include/raft/distance/sparse_l2_nn.cuh

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __SPARSE_L2_NN_H
+#define __SPARSE_L2_NN_H
+
+#pragma once
+
+#include <cub/cub.cuh>
+#include <limits>
+#include <raft/distance/detail/sparse_l2_nn.cuh>
+#include <raft/distance/fused_l2_nn.cuh>
+#include <raft/handle.hpp>
+#include <raft/util/cuda_utils.cuh>
+#include <stdint.h>
+
+namespace raft {
+namespace distance {
+
+/**
+ * @brief Sparse L2 distance and 1-nearest-neighbor computation in a single call.
+ *
+ * The benefits of such a call are 2-fold: 1) eliminate the need for an
+ * intermediate buffer to store the output of gemm 2) reduce the memory read
+ * traffic on this intermediate buffer, otherwise needed during the reduction
+ * phase for 1-NN.
+ *
+ * @tparam DataT     data type
+ * @tparam OutT      output type to either store 1-NN indices and their minimum
+ *                   distances or store only the min distances. Accordingly, one
+ *                   has to pass an appropriate `ReduceOpT`
+ * @tparam IdxT      indexing arithmetic type
+ * @tparam ReduceOpT A struct to perform the final needed reduction operation
+ *                   and also to initialize the output array elements with the
+ *                   appropriate initial value needed for reduction.
+ *
+ * @param[out] min           will contain the reduced output (Length = `m`)
```

Comment:
The order of arguments should be: `handle, params_struct, in, out, additional args`. The params_struct is optional for those algorithms that have a lot of parameters. 

Could we include a variant which a allows us to specify separate outputs for index and distance arrays? Ideally we would be able to have both, but this is something that has caused some problems when we want them separated and then need to take up additional memory to write the keyvaluepair output to additional arrays.

Reply:
All done, apart from this one:

> Could we include a variant which a allows us to specify separate outputs for index and distance arrays?

This would be possible. Maybe we want to standardize on an interface between fused_l2_nn and masked_l2_nn?

Reply:
> This would be possible. Maybe we want to standardize on an interface between fused_l2_nn and masked_l2_nn

That sounds like a great idea. And I think there's uses for having both the array<pair> and (array, array) cases.